### PR TITLE
GDScript: Fix some lambda bugs

### LINKED ISSUE #69491: GDScript 2.0: Lambda function seemingly not being called when static method of class with inferred type is called inside
### Godot version

v4.0.beta7.official [0bb1e89fb]

### System information

macOS Monterey 12.6.1

### Issue description

When assigning a class (local to script or global) to a local variable with inferred typing, then calls to static methods of that class using the variable inside a lambda function which is called by the static method will not call the lambda function. Using an instance of the class works. Also when the parameter of the method being called has not type `Callable`, the editor crashes.

### Steps to reproduce

The script below only prints

```
1
3
4
```

`2` is missing.

Can be run inside editor:

```gdscript
@tool
extends EditorScript


func _run() -> void:
	# inferring variable type with :=
	var inferred_type := TestClass

	# works
	inferred_type.call_fn(func () -> void:
		print(1)
	)

	# this crashes the editor
	#inferred_type.call_fn_no_type(func () -> void:
	#	print("crash")
	#	inferred_type.test()
	#)

	# does not get called for variable with inferred type and static method call to class
	inferred_type.call_fn(func () -> void:
		print(2)
		# works when removing this line
		inferred_type.test()
	)

	# no explicit type set
	var no_type = TestClass

	# works for variable without type
	no_type.call_fn(func () -> void:
		print(3)
		no_type.test()
	)

	# instance with inferred type
	var instance := TestClass.new()

	# works for instances
	instance.call_fn(func () -> void:
		print(4)
		instance.test()
	)


class TestClass:

	# parameter with type
	static func call_fn(fn: Callable) -> void:
		fn.call()

	# parameter without type
	static func call_fn_no_type(fn) -> void:
		fn.call()

	static func test() -> void:
		pass
```

### Minimal reproduction project

[issue.zip](https://github.com/godotengine/godot/files/10145978/issue.zip)

### LINKED ISSUE #79707: Crash when lambda connected to signal attempts to use freed instance
### Godot version

v4.1.2.rc.custom_build [6884be6b1]

### System information

Godot v4.1.2.rc (6884be6b1) - Debian GNU/Linux trixie/sid trixie - Vulkan (Mobile) - dedicated NVIDIA GeForce GTX 1060 (nvidia) - Intel(R) Core(TM) i5-8400 CPU @ 2.80GHz (6 Threads)

### Issue description

When:
1. A lambda is connected to a signal
2. The lambda references some node
3. That node gets freed
4. The signal is then triggered

The game will crash. If running in the editor, no feedback is produced in the output or debugger.

I initially saw this on `v4.1.stable.official [970459615]`, so compiled the engine from the current `4.1` tag to get a backtrace with debug symbols:

```
Godot Engine v4.1.2.rc.custom_build.6884be6b1 - https://godotengine.org
Vulkan API 1.3.224 - Forward+ - Using Vulkan Device #1: NVIDIA - NVIDIA GeForce GTX 1060
 
Connecting
Freeing
Emitting
ERROR: Can't get method on CallableCustom "<anonymous lambda>(lambda)".
   at: get_method (core/variant/callable.cpp:386)

================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.1.2.rc.custom_build (6884be6b173d22c758806e2207d64f080e9ea7c1)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x3bf90) [0x7fb6926c7f90] (??:0)
[2] Variant::get_type() const (/data/src/godot/engine/./core/variant/variant.h:330)
[3] Variant::get_call_error_text(Object*, StringName const&, Variant const**, int, Callable::CallError const&) (/data/src/godot/engine/core/variant/variant.cpp:3657 (discriminator 4))
[4] Variant::get_callable_error_text(Callable const&, Variant const**, int, Callable::CallError const&) (/data/src/godot/engine/core/variant/variant.cpp:3692)
[5] Object::emit_signalp(StringName const&, Variant const**, int) (/data/src/godot/engine/core/object/object.cpp:1082 (discriminator 4))
[6] Node::emit_signalp(StringName const&, Variant const**, int) (/data/src/godot/engine/scene/main/node.cpp:3571)
[7] Signal::emit(Variant const**, int) const (/data/src/godot/engine/core/variant/callable.cpp:463)
[8] _VariantCall::func_Signal_emit(Variant*, Variant const**, int, Variant&, Callable::CallError&) (/data/src/godot/engine/core/variant/variant_call.cpp:1061)
[9] bin/godot.linuxbsd.editor.dev.x86_64(+0x724f64a) [0x55e9222aa64a] (/data/src/godot/engine/core/variant/variant_call.cpp:2061)
[10] Variant::callp(StringName const&, Variant const**, int, Variant&, Callable::CallError&) (/data/src/godot/engine/core/variant/variant_call.cpp:1188)
[11] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Callable::CallError&, GDScriptFunction::CallState*) (/data/src/godot/engine/modules/gdscript/gdscript_vm.cpp:1690)
[12] GDScriptFunctionState::resume(Variant const&) (/data/src/godot/engine/modules/gdscript/gdscript_function.cpp:252)
[13] GDScriptFunctionState::_signal_callback(Variant const**, int, Callable::CallError&) (/data/src/godot/engine/modules/gdscript/gdscript_function.cpp:203)
[14] MethodBindVarArgTR<GDScriptFunctionState, Variant>::call(Object*, Variant const**, int, Callable::CallError&) const (/data/src/godot/engine/./core/object/method_bind.h:261 (discriminator 4))
[15] Object::callp(StringName const&, Variant const**, int, Callable::CallError&) (/data/src/godot/engine/core/object/object.cpp:739 (discriminator 2))
[16] Callable::callp(Variant const**, int, Variant&, Callable::CallError&) const (/data/src/godot/engine/core/variant/callable.cpp:62)
[17] CallableCustomBind::call(Variant const**, int, Variant&, Callable::CallError&) const (/data/src/godot/engine/core/variant/callable_bind.cpp:145)
[18] Callable::callp(Variant const**, int, Variant&, Callable::CallError&) const (/data/src/godot/engine/core/variant/callable.cpp:64)
[19] Object::emit_signalp(StringName const&, Variant const**, int) (/data/src/godot/engine/core/object/object.cpp:1071)
[20] Error Object::emit_signal<>(StringName const&) (/data/src/godot/engine/./core/object/object.h:881)
[21] SceneTree::process_timers(double, bool) (/data/src/godot/engine/scene/main/scene_tree.cpp:580 (discriminator 4))
[22] SceneTree::process(double) (/data/src/godot/engine/scene/main/scene_tree.cpp:520)
[23] Main::iteration() (/data/src/godot/engine/main/main.cpp:3425)
[24] OS_LinuxBSD::run() (/data/src/godot/engine/platform/linuxbsd/os_linuxbsd.cpp:912)
[25] bin/godot.linuxbsd.editor.dev.x86_64(main+0x16e) [0x55e91dae0f27] (/data/src/godot/engine/platform/linuxbsd/godot_linuxbsd.cpp:76)
[26] /lib/x86_64-linux-gnu/libc.so.6(+0x2718a) [0x7fb6926b318a] (??:0)
[27] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x85) [0x7fb6926b3245] (??:0)
[28] bin/godot.linuxbsd.editor.dev.x86_64(_start+0x21) [0x55e91dae0cf1] (??:?)
-- END OF BACKTRACE --
================================================================
```

---

### Expectation

If the lambda is replaced by a regular callable (eg. `do_boop.bind(...)`), then the editor will provide helpful error messages and debugging information:

Screenshot_20230720_143236

I would expect something similar when a lambda is being used.

### Steps to reproduce

This is the script from the attached MRP. The key points appear to be:

1. The `signal` is defined in a persistent location (ie. doesn't itself get free'd)
2. A lambda is used in the `connect` call
3. The crash is triggered on the `emit()` call

```gdscript
extends Node2D

signal boop

func _ready():
	var level : Node2D = $Level
	print("Connecting")
	
	# Will show runtime errors in editor
	#boop.connect(do_boop.bind(level))
	
	# Will crash with no editor feedback
	boop.connect(func(): do_boop(level))
	
	print("Freeing")
	level.queue_free()
	await get_tree().create_timer(0.1).timeout
	print("Emitting")
	boop.emit()
	print("Done")
	

func do_boop(level):
	level.position = Vector2.ZERO
```

(Strangely, if running in the editor, the `Emitting` text does not appear in the output tab, though it does appear in STDERR)

### Minimal reproduction project

Self-contained scene which reliably reproduces the crash:

[lambda_crash.zip](https://github.com/godotengine/godot/files/12106788/lambda_crash.zip)

### LINKED ISSUE #81572: GDScript crash when calling anonymous lambda stored as Callable
### Godot version

4.2.dev (98b50eb3083094a352b36b184d7b60b77ad982fe)

### System information

Mageia 9 - Vulkan (Forward+) - dedicated AMD Radeon RX Vega M GL Graphics (RADV VEGAM) () - Intel(R) Core(TM) i7-8705G CPU @ 3.10GHz (8 Threads)

### Issue description

I'm getting a crash in a project when using Godot 4.2-dev4 and later, which worked fine in 4.2-dev2 and earlier (including 4.1.x.). With 4.2-dev3, there's no crash but another regression preventing from running the project. So the crash causing regression happened between dev2 and dev4.

Stacktrace:
```
Thread 1 "godot.linuxbsd." received signal SIGSEGV, Segmentation fault.
0x0000000005459e80 in Variant::get_type (this=0x0) at ./core/variant/variant.h:330
330                     return type;

(gdb) bt
#0  0x0000000005459e80 in Variant::get_type (this=0x0) at ./core/variant/variant.h:330
#1  0x0000000005f0847d in GDScriptFunction::_get_call_error (this=0xddf0280, p_err=..., p_where=..., argptrs=0x7fffb0003f88) at ./modules/gdscript/gdscript_vm.cpp:121
#2  0x0000000005ec0a43 in GDScriptFunction::call (this=0xddf0280, p_instance=0xff6d690, p_args=0x0, p_argcount=0, r_err=..., p_state=0xf858c70) at ./modules/gdscript/gdscript_vm.cpp:1773
#3  0x0000000005e8d74c in GDScriptFunctionState::resume (this=0xf858b00, p_arg=...) at ./modules/gdscript/gdscript_function.cpp:216
#4  0x0000000005e8cfd3 in GDScriptFunctionState::_signal_callback (this=0xf858b00, p_args=0x7fffffff04f0, p_argcount=2, r_error=...) at ./modules/gdscript/gdscript_function.cpp:166
#5  0x0000000005fb0bb1 in MethodBindVarArgTR<GDScriptFunctionState, Variant>::call (this=0xdd3ab60, p_object=0xf858b00, p_args=0x7fffffff04f0, p_arg_count=2, r_error=...) at ./core/object/method_bind.h:260
#6  0x0000000009b0bafa in Object::callp (this=0xf858b00, p_method=..., p_args=0x7fffffff04f0, p_argcount=2, r_error=...) at ./core/object/object.cpp:766
#7  0x0000000009812ee0 in Callable::callp (this=0xfe069e0, p_arguments=0x7fffffff04f0, p_argcount=2, r_return_value=..., r_call_error=...) at ./core/variant/callable.cpp:62
#8  0x00000000098163bc in CallableCustomBind::call (this=0xfe069d0, p_arguments=0x7fffffff0850, p_argcount=1, r_return_value=..., r_call_error=...) at ./core/variant/callable_bind.cpp:144
#9  0x0000000009812bfe in Callable::callp (this=0xe28f1e0, p_arguments=0x7fffffff0850, p_argcount=1, r_return_value=..., r_call_error=...) at ./core/variant/callable.cpp:50
#10 0x0000000009b0d876 in Object::emit_signalp (this=0x7fffb0005120, p_name=..., p_args=0x7fffffff0850, p_argcount=1) at ./core/object/object.cpp:1113
#11 0x0000000005f2274e in Object::emit_signal<Variant> (this=0x7fffb0005120, p_name=...) at ./core/object/object.h:899
#12 0x0000000005e8d945 in GDScriptFunctionState::resume (this=0x7fffb0005120, p_arg=...) at ./modules/gdscript/gdscript_function.cpp:237
#13 0x0000000005e8cfd3 in GDScriptFunctionState::_signal_callback (this=0x7fffb0005120, p_args=0x7fffffff0d00, p_argcount=1, r_error=...) at ./modules/gdscript/gdscript_function.cpp:166
#14 0x0000000005fb0bb1 in MethodBindVarArgTR<GDScriptFunctionState, Variant>::call (this=0xdd3ab60, p_object=0x7fffb0005120, p_args=0x7fffffff0d00, p_arg_count=1, r_error=...) at ./core/object/method_bind.h:260
#15 0x0000000009b0bafa in Object::callp (this=0x7fffb0005120, p_method=..., p_args=0x7fffffff0d00, p_argcount=1, r_error=...) at ./core/object/object.cpp:766
#16 0x0000000009812ee0 in Callable::callp (this=0xfe0b970, p_arguments=0x7fffffff0d00, p_argcount=1, r_return_value=..., r_call_error=...) at ./core/variant/callable.cpp:62
#17 0x00000000098163bc in CallableCustomBind::call (this=0xfe0b960, p_arguments=0x7fffffff12b8, p_argcount=0, r_return_value=..., r_call_error=...) at ./core/variant/callable_bind.cpp:144
#18 0x0000000009812bfe in Callable::callp (this=0xfe123c8, p_arguments=0x7fffffff12b8, p_argcount=0, r_return_value=..., r_call_error=...) at ./core/variant/callable.cpp:50
#19 0x0000000009b0d876 in Object::emit_signalp (this=0xe246e60, p_name=..., p_args=0x7fffffff12b8, p_argcount=0) at ./core/object/object.cpp:1113
#20 0x000000000981590e in Signal::emit (this=0x7fffffff1290, p_arguments=0x7fffffff12b8, p_argcount=0) at ./core/variant/callable.cpp:463
#21 0x000000000998cc60 in _VariantCall::func_Signal_emit (v=0x7fffffff1288, p_args=0x7fffffff12b8, p_argcount=0, r_ret=..., r_error=...) at ./core/variant/variant_call.cpp:1060
#22 0x000000000986b841 in Method_Signal_emit::call (base=0x7fffffff1288, p_args=0x7fffffff12b8, p_argcount=0, r_ret=..., p_defvals=..., r_error=...) at ./core/variant/variant_call.cpp:2058
#23 0x0000000009826d79 in Variant::callp (this=0x7fffffff1288, p_method=..., p_args=0x7fffffff12b8, p_argcount=0, r_ret=..., r_error=...) at ./core/variant/variant_call.cpp:1182
#24 0x0000000005ec05b5 in GDScriptFunction::call (this=0xde31900, p_instance=0xff6e430, p_args=0x7fffffff6c78, p_argcount=0, r_err=..., p_state=0x0) at ./modules/gdscript/gdscript_vm.cpp:1733
#25 0x0000000005e1b01f in GDScriptInstance::callp (this=0xff6e430, p_method=..., p_args=0x7fffffff6c78, p_argcount=0, r_error=...) at ./modules/gdscript/gdscript.cpp:1843
#26 0x0000000009b0ba13 in Object::callp (this=0xe246e60, p_method=..., p_args=0x7fffffff6c78, p_argcount=0, r_error=...) at ./core/object/object.cpp:744
#27 0x0000000009826cc0 in Variant::callp (this=0x7fffffff6b70, p_method=..., p_args=0x7fffffff6c78, p_argcount=0, r_ret=..., r_error=...) at ./core/variant/variant_call.cpp:1170
#28 0x0000000005ec05b5 in GDScriptFunction::call (this=0xde31d40, p_instance=0xff6e430, p_args=0x7fffffffc780, p_argcount=4, r_err=..., p_state=0x0) at ./modules/gdscript/gdscript_vm.cpp:1733
#29 0x0000000005e1b01f in GDScriptInstance::callp (this=0xff6e430, p_method=..., p_args=0x7fffffffc780, p_argcount=4, r_error=...) at ./modules/gdscript/gdscript.cpp:1843
#30 0x0000000009b0ba13 in Object::callp (this=0xe246e60, p_method=..., p_args=0x7fffffffc780, p_argcount=4, r_error=...) at ./core/object/object.cpp:744
#31 0x0000000009812ee0 in Callable::callp (this=0xd9ed1f0, p_arguments=0x7fffffffc780, p_argcount=4, r_return_value=..., r_call_error=...) at ./core/variant/callable.cpp:62
#32 0x0000000009b0d876 in Object::emit_signalp (this=0xdf89bb0, p_name=..., p_args=0x7fffffffc780, p_argcount=4) at ./core/object/object.cpp:1113
#33 0x0000000007bcd462 in Node::emit_signalp (this=0xdf89bb0, p_name=..., p_args=0x7fffffffc780, p_argcount=4) at ./scene/main/node.cpp:3587
#34 0x0000000007c3c62d in Object::emit_signal<int, int, Vector<String>, Vector<unsigned char> > (this=0xdf89bb0, p_name=...) at ./core/object/object.h:899
#35 0x0000000007ba9e96 in HTTPRequest::_request_done (this=0xdf89bb0, p_status=0, p_code=200, p_headers=..., p_data=...) at ./scene/main/http_request.cpp:481
#36 0x0000000007ccf671 in call_with_variant_args_helper<__UnexistingClass, int, int, Vector<String> const&, Vector<unsigned char> const&, 0ul, 1ul, 2ul, 3ul> (p_instance=0xdf89bb0, 
    p_method=(void (__UnexistingClass::*)(__UnexistingClass * const, int, int, const Vector<String> &, const Vector<unsigned char> &)) 0x7ba9e16 <HTTPRequest::_request_done(int, int, Vector<String> const&, Vector<unsigned char> const&)>, p_args=0x7fffffffc970, r_error=...) at ./core/variant/binder_common.h:303
#37 0x0000000007cb4384 in call_with_variant_args_dv<__UnexistingClass, int, int, Vector<String> const&, Vector<unsigned char> const&> (p_instance=0xdf89bb0, 
    p_method=(void (__UnexistingClass::*)(__UnexistingClass * const, int, int, const Vector<String> &, const Vector<unsigned char> &)) 0x7ba9e16 <HTTPRequest::_request_done(int, int, Vector<String> const&, Vector<unsigned char> const&)>, p_args=0x7fffffffcbe0, p_argcount=4, r_error=..., default_values=...) at ./core/variant/binder_common.h:450
#38 0x0000000007c9a990 in MethodBindT<int, int, Vector<String> const&, Vector<unsigned char> const&>::call (this=0xc755f50, p_object=0xdf89bb0, p_args=0x7fffffffcbe0, p_arg_count=4, r_error=...)
    at ./core/object/method_bind.h:331
#39 0x0000000009b0bafa in Object::callp (this=0xdf89bb0, p_method=..., p_args=0x7fffffffcbe0, p_argcount=4, r_error=...) at ./core/object/object.cpp:766
#40 0x0000000009812ee0 in Callable::callp (this=0xd3723d0, p_arguments=0x7fffffffcbe0, p_argcount=4, r_return_value=..., r_call_error=...) at ./core/variant/callable.cpp:62
#41 0x0000000009b06019 in CallQueue::_call_function (this=0xad970e0, p_callable=..., p_args=0xd3723e8, p_argcount=4, p_show_error=false) at ./core/object/message_queue.cpp:219
#42 0x0000000009b0673e in CallQueue::flush (this=0xad970e0) at ./core/object/message_queue.cpp:324
#43 0x0000000007bd0a57 in SceneTree::physics_process (this=0xd3886e0, p_time=0.016666666666666666) at ./scene/main/scene_tree.cpp:471
#44 0x000000000547e04b in Main::iteration () at main/main.cpp:3467
#45 0x00000000053fd320 in OS_LinuxBSD::run (this=0x7fffffffcf50) at platform/linuxbsd/os_linuxbsd.cpp:933
#46 0x00000000053f5dc0 in main (argc=1, argv=0x7fffffffd538) at platform/linuxbsd/godot_linuxbsd.cpp:74
```

This works around the crash, but the call still fails so functionality is broken:
```diff
diff --git a/modules/gdscript/gdscript_vm.cpp b/modules/gdscript/gdscript_vm.cpp
index a7dc0b6d59..7545108c53 100644
--- a/modules/gdscript/gdscript_vm.cpp
+++ b/modules/gdscript/gdscript_vm.cpp
@@ -116,6 +116,7 @@ String GDScriptFunction::_get_call_error(const Callable::CallError &p_err, const
 
 	if (p_err.error == Callable::CallError::CALL_ERROR_INVALID_ARGUMENT) {
 		int errorarg = p_err.argument;
+		ERR_FAIL_NULL_V_MSG(argptrs[errorarg], "Invalid argument type in " + p_where + ".", "CallError holds invalid pointers to arguments.");
 		// Handle the Object to Object case separately as we don't have further class details.
 #ifdef DEBUG_ENABLED
 		if (p_err.expected == Variant::OBJECT && argptrs[errorarg]->get_type() == p_err.expected) {
```
Might be worth committing anyway to catch such bugs if we prefer error prints to a nullptr dereference.

With this patch, I get this output:
```
ERROR: CallError holds invalid pointers to arguments.
   at: _get_call_error (./modules/gdscript/gdscript_vm.cpp:119)
SCRIPT ERROR: Invalid argument type in function '<anonymous lambda>(lambda) (Callable)'.
          at: async (res://addons/w4gd/rest-client/client_promise.gd:44)
SCRIPT ERROR: Invalid call. Nonexistent function 'is_error' in base 'Nil'.
          at: host_game_w4_cloud (res://autoloads/Net.gd:43)
```
The script errors are the consequence of `await`'ing a `Callable.call()` that fails.

---

I'm seeing this on an internal W4 demo, which isn't public yet.

I haven't attempted to extract a MRP yet, but I've done some checks to identify what caused the regression. As pointed out above, it's tricky to bisect as there have been multiple regressions between dev2 and dev4.

But I have a strong suspicion for #79880 or one of its follow-up regression fixes. I confirmed it by reverting these 4 related commits (first 3 needed to revert #79880 cleanly), which solves the bug.

```
commit a26cffb2c473de7875298989301f3b95b3fa9498 (HEAD -> test)
Author: Rémi Verschelde <rverschelde@gmail.com>
Date:   Tue Sep 12 13:12:54 2023 +0200

    Revert "GDScript: Fix bug with identifier shadowed below in current scope"
    
    This reverts commit d53fc92b4c6b5e4484e8f0bfff6ac55163dde3fb.

commit 880735f89bdc740fb7c38dce162c4f900e46133f
Author: Rémi Verschelde <rverschelde@gmail.com>
Date:   Tue Sep 12 13:12:49 2023 +0200

    Revert "GDScript: Allow use local constants as types"
    
    This reverts commit 68a567bd1389a2cb410fc002632ccd5b5fb59f5c.

commit efd2eda69d538fb495df195df6a5309f19e76dfc
Author: Rémi Verschelde <rverschelde@gmail.com>
Date:   Tue Sep 12 13:06:38 2023 +0200

    Revert "GDScript: Fix regression with native signal not found"
    
    This reverts commit f19377160cb6cee6db223660ce8df02d18c38a08.

commit 27f5f22ec31998af9dc960e7158e863fdb417bfc
Author: Rémi Verschelde <rverschelde@gmail.com>
Date:   Tue Sep 12 13:06:23 2023 +0200

    Revert "GDScript: Fix "Identifier not found" error when accessing inner class from inside"

    This reverts commit fb45cab133103de4f641cc72a553a28463bed7fd.
```

*Edit:* I've tested with revert just the 3 last PRs and not #79980, and I get an endless stream of compilation errors, like in 4.2-dev3. So the problem did start with #79980, and morphed from compilation errors to a lambda call failure in a follow-up.

CC @dalexeev 

### Steps to reproduce

TBD

### Minimal reproduction project

The project where I reproduced it is moderately complex and I haven't extracted a MRP yet. I can't share the project publicly yet, but I can share it privately with GDScript maintainers.

### PR DESCRIPTION
**Pull request by dalexeev**
* Closes #69491.
* Closes #79707.
* Closes #81572.

As stated in #81572, `GDScriptFunction::_get_call_error()` can crash the engine.

https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/modules/gdscript/gdscript_vm.cpp#L118-L121

https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/modules/gdscript/gdscript_vm.cpp#L1773-L1773

I decided to check what `err.argument` was equal to, and it turned out to be `-1`. This is because lambda captures are hidden arguments of lambda function.

https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/modules/gdscript/gdscript_lambda_callable.cpp#L77-L88

If for some reason a capture is invalid (for example, due to a bug in the compiler or if the object has been freed at this point), then this leads to a negative argument index and a crash.

https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/modules/gdscript/gdscript_vm.cpp#L545-L548

https://github.com/godotengine/godot/blob/3ed4497113fa10611b90290ce22a751fb9d26e2e/modules/gdscript/gdscript_function.h#L109-L113

* (See also #80844.)

**Status**

* #62769
  * :no_entry_sign: This bug is difficult to reproduce. It seems to be related to hot reloading.
* #74607 
  * :no_entry_sign: Some crashes were prevented, but the hot reloading bug remained (which can also cause a crash in itself).
* #69491
  * :white_check_mark: Fixed.
* A bug with freed capture.
  * #79707
  * :white_check_mark: Fixed.
* #81572
  * :white_check_mark: Fixed ([comment](https://github.com/godotengine/godot/pull/81605#issuecomment-1717325026)). 

```gdscript
@tool
extends EditorScript

func _run() -> void:
    var x: Node = Node.new()
    x.free()
    (func (): print(x)).call()
```

### CODE REVIEW

File: modules/gdscript/gdscript_vm.cpp [hunk 1]
```diff
@@ -116,6 +116,7 @@ String GDScriptFunction::_get_call_error(const Callable::CallError &p_err, const
 
 	if (p_err.error == Callable::CallError::CALL_ERROR_INVALID_ARGUMENT) {
 		int errorarg = p_err.argument;
+		ERR_FAIL_NULL_V_MSG(argptrs[errorarg], "Invalid argument type in " + p_where + ".", "CallError holds invalid pointers to arguments.");
```
Comment (dalexeev) [inline modules/gdscript/gdscript_vm.cpp:119]:
```suggestion
		ERR_FAIL_COND_V_MSG(errorarg < 0 || argptrs[errorarg] == nullptr, "Invalid CallError argument index or pointers in " + p_where + ".", "Invalid CallError argument index or pointers.");
```

Comment (akien-mga) [reply modules/gdscript/gdscript_vm.cpp:119]: The rephrasing for the `MSG` seems good.

The reason I used the more generic "Invalid argument type in ..." message for the return value is that this is the type of call error we're handling here. The return value should be a message that relates to the bug in the user's script that caused the call error. And the `ERR_FAIL` message is for us developers.

**Comment by akien-mga** [comment]
I confirm this fixes https://github.com/godotengine/godot/issues/81572 on my project :tada: 

I tested #62769 and documented my findings there for the current `master` branch. This PR doesn't seem to improve or worsen things.

**Comment by dalexeev** [comment]
Thanks for testing and documenting! I also saw some of these and several others. It looks like hot reloading can cause a random effect, depending on how the bytecode length (or something else) changes. In theory this can be dangerous.

**Comment by dalexeev** [comment]
Maybe this also fixes/removes a crash #79707, #75421 and/or #76108. I'll test it later.

File: modules/gdscript/gdscript_vm.cpp (hunk 1 above)
Comment (akien-mga) [reply modules/gdscript/gdscript_vm.cpp:119]: That being said if we consider this condition should never be hit, and thus doesn't need to give useful info about potential script errors, we can just go with something like "GDScript bug, please report" and the `MSG` giving details.

Comment (dalexeev) [reply modules/gdscript/gdscript_vm.cpp]: Done.

**Review by akien-mga — APPROVED** [review]
I'm not proficient with the GDScriptCompiler but I tested the changes and confirmed the bug fixes, without visible regression in a project that uses lambdas and awaits them a lot.

So I think we can merge so this gets tested in dev5.

**Review by adamscott — APPROVED** [review]
Another great PR from our @dalexeev that pass the test! Thanks for the numerous tests added.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**